# SAMPLES = ["sample.R1", "sample.R2"]

## TODO later: What if the fastqc report shows bad results?
## How can we make the pipeline fail if the quality report is bad?
rule fastqc:
    input:
        "{sample}.fastq.gz"
    output:
        html="qc/fastqc/{sample}.html",
        zip="qc/fastqc/{sample}_fastqc.zip" # the suffix _fastqc.zip is
        # necessary for multiqc to find the file. If not using multiqc, you are
        # free to choose an arbitrary filename
    params: "--quiet"
    log:
        "logs/fastqc/{sample}.log"
    threads: 1
    wrapper:
        "0.77.0/bio/fastqc"


rule trimmomatic_pe:
    input:
        r1="{sample}.R1.fastq.gz",
        r2="{sample}.R2.fastq.gz"
    output:
        r1="trimmed/{sample}.R1.fastq.gz",
        r2="trimmed/{sample}.R2.fastq.gz",
        # reads where trimming entirely removed the mate
        r1_unpaired="trimmed/{sample}.R1.unpaired.fastq.gz",
        r2_unpaired="trimmed/{sample}.R2.unpaired.fastq.gz"
    log:
        "logs/trimmomatic/{sample}.log"
    params:
        # list of trimmers (see manual)
        trimmer=["TRAILING:3", "LEADING:20", "TRAILING:20",
        "SLIDINGWINDOW:4:30", "MINLEN:36", "HEADCROP:0",
        "ILLUMINACLIP:/usr/local/programs/Trimmomatic-0.36/adapters/TruSeq3-PE.fa:2:30:10"],
        # optional parameters
        extra="",
        compression_level="-9"
    threads:
        1
    # optional specification of memory usage of the JVM that snakemake will respect with global
    # resource restrictions (https://snakemake.readthedocs.io/en/latest/snakefiles/rules.html#resources)
    # and which can be used to request RAM during cluster job submission as `{resources.mem_mb}`:
    # https://snakemake.readthedocs.io/en/latest/executing/cluster.html#job-properties
    resources:
        mem_mb=1024
    wrapper:
        "0.77.0/bio/trimmomatic/pe"
#
#
#
# rule bwa_map:
#     input:
#         "data/genome.fa",
#         "data/samples/{sample}.fastq"
#     output:
#         "mapped_reads/{sample}.bam"
#     shell:
#         "bwa mem {input} | samtools view -Sb - > {output}"
#
#
